proposals/utils/proposal_utils.py

```diff
@@ -373,3 +377,60 @@ def notify_local_staff(proposal):
 
     # Reset the current language
     activate(current_language)
+
+
+@deconstructible
+class FilenameFactory:
+    '''A callable class which can be passed to upload_to() in FileFields
+    and can be deconstructed for migrations'''
+    
+    def __init__(self, document_type):
+        self.document_type = document_type
+    
+    def __call__(self, instance, original_fn):
+        '''Returns a custom filename preserving the original extension,
+        something like "FETC-2020-002-01-Villeneuve-T2-Informed-Consent.pdf"'''
+        
+        # Importing here to prevent circular import
+        from proposals.models import Proposal
+        
+        if isinstance(instance, Proposal):
+            proposal = instance
+            trajectory = None
+        else:
+            # In case of Documents or Study objects
+            proposal = instance.proposal
+            try:
+                trajectory = 'T' + str(instance.study.id)
+            except AttributeError:
+                trajectory = None
+        
+        lastname = proposal.created_by.last_name
+        refnum = proposal.reference_number
+        
+        extension = '.' + original_fn.split('.')[-1][-7:] # At most 7 chars seems reasonable
+        
+        fn_parts = [ p for p in ['FETC',
+                                 refnum,
+                                 lastname,
+                                 trajectory,
+                                 self.document_type,
+                                 ] if p != None or '' ]
```

Comment:
Isn't it better to filter out None's instead?

Reply:
Some Django fields will return an empty string yet not be equivalent to None. I have now used filter() because it's more obvious than a comprehension.